Enable instantiating model with pretrained backbone weights

amyeroberts opened the pull request:
# What does this PR do?

At the moment, there's two ways backbone weights can be initialized in a model: 
* Pretrained weights that are already saved alongside the parent model's weights in a checkpoint
* Randomly initialized from config

However, the main use-case of backbones is to be able to leverage pretrained weights of a feature extractor which then feed features to the neck/head. It's therefore necessary to be able instantiate a new model e.g. a MaskFormer model with pretrained backbone weights and the rest of the weights randomly initialized. 

This PR adds a new function `load_backbone` which enables loading a backbone from either a backbone config or using the parent model's config which contains information about the backbone e.g. the checkpoint. 

To reduce the number of changes, this PR doesn't update the modeling code to use `load_backbone`. Once this PR is added and modeling code updated, it will be possible to load backbones in a variety of ways:

Initialize a new MaskFormerModel with a pretrained resnet backbone
```py
from transformers import MaskFormerConfig, MaskFormerForInstanceSegmentation, ResNetConfig

config = MaskFormerConfig(backbone="microsoft/resnet50", use_pretrained_backbone=True)
model = MaskFormerForInstanceSegmentation(config)
```

Initialize a new MaskFormerModel with a randomly initalized resnet backbone
```py
from transformers import MaskFormerConfig, MaskFormerForInstanceSegmentation, ResNetConfig

config = MaskFormerConfig(backbone="microsoft/resnet50", use_pretrained_backbone=False)
model = MaskFormerForInstanceSegmentation(config)
```

Initialize a new MaskFormerModel with a backbone config
```py
backbone_config = ResNetConfig()
config = MaskFormerConfig(backbone_config=backbone_config)
model = MaskFormerForInstanceSegmentation(config)
```


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

amyeroberts committed:
Enable instantiating model with pretrained backbone weights

amyeroberts commented on src/transformers/models/vitmatte/modeling_vitmatte.py:
```diff
@@ -80,6 +80,17 @@ class VitMattePreTrainedModel(PreTrainedModel):
     supports_gradient_checkpointing = True
 
     def _init_weights(self, module):
+        # # If the module is the backbone, then use the backbone's init weights function
+        # print(module.__class__)
+
+        # if hasattr(module, "backbone_type"):
+        #     # print(module.__class__)
+        #     pass
+        #     # import pdb; pdb.set_trace()
+        #     # print("here", module)
+
+        # # if hasattr(module, "backbone_type"):
+        # #     module._init_weights()
```
```suggestion
```

amyeroberts commented on src/transformers/models/maskformer/modeling_maskformer.py:
```diff
@@ -23,7 +23,6 @@
 from torch import Tensor, nn
 
-from ... import AutoBackbone
 from ...activations import ACT2FN
```
```suggestion
from ... import AutoBackbone
from ...activations import ACT2FN
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28214). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

amyeroberts committed:
Update tests so backbone checkpoint isn't passed in

amyeroberts committed:
Remove doc updates until changes made in modeling code

amyeroberts commented:
cc @NielsRogge

ArthurZucker commented on src/transformers/models/conditional_detr/configuration_conditional_detr.py:
```diff
@@ -93,11 +93,12 @@ class ConditionalDetrConfig(PretrainedConfig):
-            Name of convolutional backbone to use in case `use_timm_backbone` = `True`. Supports any convolutional
-            backbone from the timm package. For a list of all available models, see [this
-            page](https://rwightman.github.io/pytorch-image-models/#load-a-pretrained-model).
+            Name of backbone to use when `backbone_config` is `None`. If `use_pretrained_backbone` is `True`, this
+            will load the corresponding pretrained weights from the timm or transformers library. If `use_pretrained_backbone`
+            is `False`, this loads the backbone's config and uses that to initialize the backbone with random weights.
```
both doc changes imply that `use_timm_backbone` is no longer a requirement for this. Just wondering if this works when `use_timm_backbone=False`

amyeroberts replied:
Yes! That's the main benefit of this PR. Previously, only timm models could be loaded as pretrained weights but the previous constraint is just a hangover of how detr models were implemented before we properly had backbone and how Autobackbone is currently used. 

However, not all of the arguments are fully compatible yet. This PR enables loading in the pretrained backbones, but most models will only use `AutoBackbone.from_config` or timm's `create_model` e.g. [here](https://github.com/huggingface/transformers/blob/45b1dfa342f2ec00634c31f635bded6e346c98d2/src/transformers/models/conditional_detr/modeling_conditional_detr.py#L366) and [here](https://github.com/huggingface/transformers/blob/45b1dfa342f2ec00634c31f635bded6e346c98d2/src/transformers/models/conditional_detr/modeling_conditional_detr.py#L352) for conditional detr. 

So, `use_pretrained_backbone=True`, `use_timm_backbone=False` are valid, but won't work until the follow up PR to use `load_backbone`. Which admiteddly is not great behaviour. I can add a validation check for this combo until it's fully enabled?

ArthurZucker replied:
yep sounds good!

amyeroberts replied:
Done!  https://github.com/huggingface/transformers/pull/28214/commits/5e565e32a41f7162fbc96d41441faaf4127f0667

ArthurZucker commented on src/transformers/models/conditional_detr/configuration_conditional_detr.py:
```diff
@@ -93,11 +93,12 @@ class ConditionalDetrConfig(PretrainedConfig):
+            is `False`, this loads the backbone's config and uses that to initialize the backbone with random weights.
         use_pretrained_backbone (`bool`, *optional*, defaults to `True`):
-            Whether to use pretrained weights for the backbone. Only supported when `use_timm_backbone` = `True`.
+            Whether to use pretrained weights for the backbone. You cannot specify both `backbone` and `backbone_config`
+            when this is `False`.
```
```suggestion
 Whether to use pretrained weights for the backbone.
```
If this is set to `True` I don't get how you can set backbone and backbone config ? (unless we try to load whatever we can from the backbone (weights) in a backbone (arch) defined by the config ? (the check seems to also ignore this arg

amyeroberts replied:
Yes, sorry, the doc is outdated vs. the check.

amyeroberts replied:
Updated in https://github.com/huggingface/transformers/pull/28214/commits/5e565e32a41f7162fbc96d41441faaf4127f0667

ArthurZucker commented on src/transformers/models/deformable_detr/configuration_deformable_detr.py:
```diff
@@ -85,11 +85,12 @@ class DeformableDetrConfig(PretrainedConfig):
+            is `False`, this loads the backbone's config and uses that to initialize the backbone with random weights.
         use_pretrained_backbone (`bool`, *optional*, defaults to `True`):
-            Whether to use pretrained weights for the backbone. Only supported when `use_timm_backbone` = `True`.
+            Whether to use pretrained weights for the backbone. You cannot specify both `backbone` and `backbone_config`
```
same comment here

ArthurZucker commented on src/transformers/utils/backbone_utils.py:
```diff
@@ -286,3 +286,46 @@ def to_dict(self):
+
+    # If there is a backbone_config and a backbone checkpoint, and use_pretrained_backbone=False then the desired
+    # behaviour is ill-defined: do you want to load from the checkpoint's config or the backbone_config?
+    if backbone_config is not None and backbone_checkpoint is not None and use_pretrained_backbone is not None:
+        raise ValueError("Cannot specify both config.backbone_config and config.backbone")
```
that's the only place where the check depends on `use_pretrained_backbone`

ArthurZucker commented on src/transformers/utils/backbone_utils.py:
```diff
@@ -286,3 +286,46 @@ def to_dict(self):
+        raise ValueError("Cannot specify both config.backbone_config and config.backbone")
+
+    if backbone_config is not None or use_timm_backbone or backbone_checkpoint is not None:
+        # This is a config from the parent model the has a backbone. This contains the pretrained backbone checkpoint
```
```suggestion
        # This is a config from the parent model that has a backbone. This contains the pretrained backbone checkpoint
```

ArthurZucker commented on tests/utils/test_backbone_utils.py:
```diff
@@ -16,11 +16,22 @@
+if is_torch_available():
+    import torch
+
+    # from transformers import PretrainedModel
+    from transformers import BertPreTrainedModel as PretrainedModel
```
could you explain why we need this?

amyeroberts replied:
I've changed this to just use `BertPreTrainedModel` - this was just something I didn't tidy up, sorry. 

`BertPreTrainedModel` is imported for the weight init for the dummy model used in tests

ArthurZucker commented on tests/utils/test_backbone_utils.py:
```diff
@@ -126,3 +137,51 @@ def test_backbone_mixin(self):
+
+    @slow
+    @require_torch
+    def test_load_backbone_in_new_model(self):
```
The main pain point is `_fast_init` when using `from_pretrained` will not initialize layers. not 100% familiar with the interactions between timm backbone and that but interesting to test `from_pretrained` here as well IMO 😉

amyeroberts replied:
What behaviour wrt `from_pretrained` do you think should be tested here? It's somewhat separate from the added functionality: a new model with randomly initialized weights can have pretrained backbone weights (which are loaded with `from_pretrained`). We know models that have pretrained weights which contain backbones can be successfully loaded back in, with all the weights intact as this is covered in the model integration tests for e.g. detr and dpt. I could save and reload the models to make sure the weights are preserved to make sure it's still true?  

Regarding timm and its interacations in our library: I believe it should be OK. The weights for the timm model are initialized in the model's init call before `post_init` and are marked with `is_hf_initialized=True`. I can add a test here to include timm to make sure it works as expected + protect against regressions.

ArthurZucker reviewed (commented):
Thanks, this looks like a nice feature! Just 1 nit on the `use_pretrained_backbone` and a test needed for from_pretrained!

amyeroberts committed:
Clarify pretrained import

amyeroberts committed:
Update configs - docs and validation check

amyeroberts committed:
Update src/transformers/utils/backbone_utils.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

amyeroberts committed:
Clarify exception message

amyeroberts committed:
Update config init in tests

amyeroberts committed:
Add test for when use_timm_backbone=True

amyeroberts commented:
As an update - when running slow tests for DETR, the instantiated model fails on the integration tests: output logits are similar, but not exactly the same. Digging into it 🕵️

amyeroberts committed:
Small test updates

amyeroberts commented:
Good news- failures were just an artefact of running on the ubuntu machine (they fail running on main too). Running on my mac, everything's good. Just one test value had a 6th decimal place change. Updated in  [06ff4f6](https://github.com/huggingface/transformers/pull/28214/commits/06ff4f66ad618e033a4b2d0925b9924649c3f39c). 

@ArthurZucker Only outstanding thing to address is for [this comment](https://github.com/huggingface/transformers/pull/28214#discussion_r1440543327). I've added tests in `test_load_backbone_in_new_model` to include checking loading with a timm backbone too. Is there anything else you'd like me to add / cover?

ArthurZucker reviewed (approved):
Yep, looks good to me now!

amyeroberts commented:
Currently failing tests are flaky. Opened a PR #28458 to address
